Add create account endpoint

rachelekm opened the pull request:
## Overview

Create a way for unregistered users to create an account and define themselves as restaurant "owners". This PR adds a new viewset that creates a new user account on POST request and, if containing restaurant data, will create a new restaurant instance as well. The Restaurant model now has a many-to-many owner field that references users who are the "owner". A restaurant can have many owners and a user can own many restaurants.
Connects #7 , #9 

### Notes

-  The creating new user accounts issue (#9 ) was separated into backend and frontend issues, and now this PR encompasses both backend commits as well as commits from #11 as current functionality was dependent on PR #11. This PR is now superseding #11, which is to be closed.

## Testing Instructions

 * `/scripts/update`
 * `/scripts/manage test` confirm tests pass
 * `/scripts/server` and open `localhost:8033/admin/` in browser
 * Confirm owner field in api > Restaurants
 * Add user as owner and confirm changes are saved

rachelekm committed:
Create test factories

rachelekm committed:
Add test cases and update serializer

- Create test cases for RestaurantSerializer and UserSerializer specific to owner and is_owner fields
- Update serializers for test pass
- Linting and formatting

rachelekm committed:
Add RestaurantAdmin

rachelekm committed:
initial test commit

rachelekm committed:
Add .venv to gitignore

rachelekm committed:
Create new user and test cases

rachelekm committed:
Update tests and UserSerializer for owner request.data

rachelekm committed:
Resolve loc error in RestaurantSerializer

rachelekm committed:
Add custom create to RestaurantSerializer

- Custom create to have writable nested field "owner" created and added without creating duplicate user instances
- Delete is_owner field in UserSerializer to stop cyclical many-to-many fields
- Modify tests for serializer field changes
- formatting/linting

rachelekm committed:
Delete .vscode/settings

rachelekm committed:
Add owner field and format serializers

- Add owner field to Restaurant
- Refactor tests to account for new owner field
- Refactor serializers for merged Restaurant fields
- Formatting & linting

ddohler commented on django/api/factories.py:
```diff
@@ -23,3 +23,14 @@ class Meta:
+
+    @factory.post_generation
+    def owner(self, create, extracted, **kwargs):
+        if not create or not extracted:
```
Usually it's good practice to short-circuit early, so I like the way you're thinking about this, but in this case I think this can be removed: the only case where this function will do anything is if `extracted` is True, so the if-statement below is enough on its own.

ddohler commented on django/api/migrations/0005_restaurant_owner.py:
```diff
@@ -1,5 +1,4 @@
-# Generated by Django 3.2.4 on 2021-08-24 18:27
```
I don't think anything needs to change here, and I know this was a result of the development process you went through for this PR, but I don't want to miss an opportunity to note that in a project with multiple developers, editing a migration file would be a warning flag because it could cause conflicts for developers (or deployed versions of the application) that had already applied the older version of the migration. But this is fine as-is! I think this was most expeditious way to achieve what you needed in this case.

ddohler commented on django/api/views.py:
```diff
@@ -1,21 +1,51 @@
 
+
+class UserViewSet(mixins.RetrieveModelMixin, mixins.UpdateModelMixin, GenericViewSet):
+    permission_classes = (IsAuthenticatedOrCreate,)
```
This works, and I don't think it needs to change, but for your consideration, another way to structure this might be to split the viewset into two pieces:
- A `CreateUserView` that only allows creation requests, and requires no authentication
- The `UserViewSet` which allows everything else and does require authentication.

But I think this is fine unless that alternative seems obviously better to you for some reason!

ddohler commented on django/api/serializers.py:
```diff
@@ -2,24 +2,59 @@
         write_only_fields = ("password",)
-        read_only_fields = ["id", "username"]
+
+    def to_internal_value(self, data):
```
I think there might be some cleanup that could be done here, but I'm not quite sure what the best change to make is. What I'm noticing is that both this serializer and the `RestaurantSerializer` assume that the user information will be available on the `"user"` field of the request body, but neither of them are entirely happy about that -- they both pop it out and do something else with it.

One thought I'm having is that it might be helpful to have some kind of coordinator serializer that would look something like this:

```
class CreateUserWithRestaurantSerializer(Serializer):
    user = UserSerializer(many=True)
    restaurant = RestaurantSerializer(required=False)
```

I'm curious whether you tried something like that and whether you ran into problems with it?

ddohler commented on django/api/serializers.py:
```diff
@@ -2,24 +2,59 @@
     """A class to serialize restaurants as GeoJSON compatible data"""
 
-    average_rating = serializers.FloatField()
+    average_rating = serializers.FloatField(required=False, allow_null=True)
```
No need to make changes here, but to flesh out the discussion we had on Slack, an alternate way to structure this would be as a `RestaurantReadSerializer` and a `RestaurantWriteSerializer`.

The `RestaurantReadSerializer` might look something like this:

```python
class RestaurantReadSerializer(GeoFeatureModelSerializer):
    average_rating = serializers.FloatField()
    class Meta:
        model = Restaurant
        geo_field = "loc"
    # No to_internal_value() or create() needed
```

while the `RestaurantWriteSerializer` would omit the `average_rating` field but would define `to_internal_value` and `create()`.

ddohler commented on django/api/tests.py:
```diff
@@ -77,3 +79,114 @@ def test_restaurants_default_order_by_name(self):
+
+class UserViewsetTestCase(TestCase):
+    def setUp(self):
+        self.c = Client()
```
```suggestion
        self.client = Client()
```

I see the motivation for wanting to keep this short because it gets used a lot, but I think the benefit of being clear about what `self.client` is for a reader coming along later is going to outweigh the benefit from faster typing. I think the best uses for one-character variable names are within small, tight loops, ideally where there is some convention or surrounding context immediately available to make the meaning clear. For example `[p.transformTo('4326') for p in polygon.points()]`.

ddohler commented on django/api/tests.py:
```diff
@@ -77,3 +79,114 @@ def test_restaurants_default_order_by_name(self):
+class UserViewsetTestCase(TestCase):
+    def setUp(self):
+        self.c = Client()
+        self.userdata_owner_1 = {
```
I would suggest moving these each as local variables into the tests that use them. Putting this type of stuff into `setUp` is seductive because it feels like the right place to put "getting ready to test" things, but I've found that as tests are added, putting everything on `self` risks creating dependencies between tests that may be difficult to untangle later on because it will be easy for variables on `self` to be used by multiple tests. I've found that trying to keep tests as self-contained as possible, even at the expense of a bit more duplication than we might accept in application code, usually results in tests that are easier to maintain long-term.

ddohler commented on django/api/tests.py:
```diff
@@ -77,3 +79,114 @@ def test_restaurants_default_order_by_name(self):
+            "/api/user/", self.userdata_not_owner, content_type="application/json"
+        )
+        # check created status
+        self.assertEqual(response.status_code, 201)
```
In these types of cases I try to remember to pass the response body as the [third parameter to `assertEqual`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual); that'll give you more info if the test fails. The way it is now, if the test fails you'll just get something like `400 != 201` but if you pass the response body as the message, you'll get whatever info the API provided about the reason for the error, something like `400 != 201; { telephone_number: "This field is required"}`

ddohler commented on django/api/tests.py:
```diff
@@ -77,3 +79,114 @@ def test_restaurants_default_order_by_name(self):
+        does_user_exist = User.objects.filter(
+            username=self.userdata_not_owner["user"]["username"]
+        ).exists()
+        self.assertEqual(True, does_user_exist)
```
```suggestion
        self.assertTrue(does_user_exist)
```

(there are a few other places where this suggestion applies as well)

ddohler replied:
You could even do `self.assertTrue(User.objects.filter(...).exists())` if you want.

ddohler commented on django/api/tests.py:
```diff
@@ -77,3 +79,114 @@ def test_restaurants_default_order_by_name(self):
+        restaurant = Restaurant.objects.get(
+            name=self.userdata_owner_2["restaurant_name"]
+        )
+        restaurant_owner = dict(
```
I'm not sure that it's necessary to re-serialize the objects you get from the database in this way before testing. It makes sense that you are posting to the API, and then querying the database via a different route to verify that the created objects match what you expect, but I think it should be sufficient to simply compare the `.pk` property of each candidate `User` object; if the primary keys match then you know you've got the same object. But I might not be understanding what's happening here correctly--what do you think?

ddohler commented on django/api/serializers.py:
```diff
@@ -2,24 +2,59 @@
+        restaurant_data = {
+            "name": data["restaurant_name"],
+            "address": data["restaurant_address"],
+            "loc": Point(lat, lng),
```
I think you might be able to avoid this if you declare `loc = GeometryField()` (`GeometryField` is available in rest_framework_gis.serializers`). Did you try that and run into trouble?

ddohler commented on django/api/serializers.py:
```diff
@@ -2,24 +2,59 @@
+
+    def to_internal_value(self, data):
+        """Format request data into correct serializer fields"""
+        user_data = data.pop("user")
```
I'm a little confused about what's happening here -- I think ideally we'd want to have the user_data in the right place on the request object from the get-go. Is this something related to switching between `UserSerializer` and `RestaurantSerializer`?

ddohler commented on django/api/views.py:
```diff
@@ -1,21 +1,51 @@
         return self.request.user
 
+    def get_serializer_class(self):
+        if self.request.data.__contains__("restaurant_name"):
```
Also no change necessary here, but as I'm looking at this and some of the hoops that DRF is making you jump through in order for it to work, I'm wondering whether making the switch on the frontend might eliminate some of them. You'd have to define two endpoints, and it's definitely not as RESTy, because you'd have something like `/api/User/create-with-restaurant/`, but making the two pieces of functionality very, very separate might allow you to have two simple logic paths to follow on the backend rather than one more complex one.

ddohler reviewed (commented):
I haven't tested yet but I had what felt like plenty of comments for you to chew on 🙂 so I decided to go ahead and submit this now, and then I can take another pass with testing after you've made any changes you want to make after reading this. Overall it's looking really good -- most of my comments were related to ways to potentially structure things to make life easier for yourself.

rachelekm committed:
Simplify serializers used in UserViewSet

- Create combined CreateUserWithRestaurantSerializer for a simpler way to serialize request
- Split restaurant serializer into a separate RestaurantReadSerializer and RestaurantWriteSerializer to avoid issues with serializing non-required average_rating and owner fields at create
- Change format of request JSON data in tests to better accommodate backend and get rid of to_internal_value methods

rachelekm committed:
Reformat tests and factories

rachelekm committed:
Remove UserViewSet custom create

No longer need to extract user data from serializer data - a fix to create consistent JSON responses when using the RestaurantSerializer. Now JSON response data is more easily accessible with better request formatting and new CreateUserWithRestaurantSerializer, so CreateModelMixin covers needs

rachelekm committed:
Separate permissions in UserViewSet

ddohler commented on django/api/tests.py:
```diff
@@ -57,23 +59,131 @@ def setUp(self):
+        self.assertEqual(response.data["owner"]["id"], restaurant.owner.first().pk)
+
+
+class RestaurantOwnerSerializerTestCase(TestCase):
```
These tests are looking great!

ddohler commented on django/api/views.py:
```diff
@@ -1,27 +1,43 @@
+        return UserSerializer
+
+#Interact with User with permissions
+class UserViewSet(mixins.RetrieveModelMixin, mixins.UpdateModelMixin, GenericViewSet):
```
Nice use of multiple inheritance here. I would probably put the `GenericViewSet` first in the list -- the order does matter for multiple inheritance although in this case I don't think it'll make a difference.

ddohler reviewed (approved):
This is looking good! I ran into trouble when setting up the VM to try to test and it didn't seem worthwhile to spend a lot of time trying to fix it--as long as it works for you that's good enough. Here's the error I saw; have you run into anything like this? The error message looks familiar but I don't remember ever having fixed something like this.

```
    default: $ react-scripts build
    default: Failed to compile.
    default: ENOTSUP: operation not supported on socket, copyfile '/usr/local/src/client/public/favicon.ico' -> '/usr/local/src/client/build/favicon.ico'
    default: error Command failed with exit code 1.
```

In any event, this is ready to go!

rachelekm commented:
Huh, I haven't received that error before. I ran `vagrant destroy` and then `scripts/setup` again just to see and didn't bump into it. It looks like it's a remnant of `create-react-app` so might be worth just deleting if it's causing problems.
